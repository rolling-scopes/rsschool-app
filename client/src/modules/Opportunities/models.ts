import { ResumeDtoEnglishLevelEnum, ResumeDtoMilitaryServiceEnum } from 'api';
import { Dayjs } from 'dayjs';

export interface CourseData {
  locationName: string;
  courseFullName: string;
  isExpelled: boolean;
  certificateId: string | null;
  isCourseCompleted: boolean;
  totalScore: number;
  rank: number | null;
  mentor: {
    githubId: string;
    name: string;
  };
}

export type VisibleCourses = number[];

export interface CourseDataShortened {
  courseId: number;
  courseFullName: string;
}

export interface VisibleCoursesFormData {
  [id: string]: boolean;
}

export type ContactType = 'phone' | 'email' | 'skype' | 'telegram' | 'linkedin' | 'githubUsername' | 'website';

export interface UserData {
  avatarLink: string | null;
  name: string | null;
  desiredPosition: string | null;
  selfIntroLink: string | null;
  englishLevel: ResumeDtoEnglishLevelEnum | null;
  militaryService: ResumeDtoMilitaryServiceEnum | null;
  notes: string | null;
  startFrom: string | null;
  fullTime: boolean;
  locations: string | null;
}

export type Contacts = {
  [key in ContactType]: string | null;
};

export interface FieldData {
  name: string[];
  value: any;
  touched: boolean;
  validating: boolean;
  errors: string[];
}

export interface AllUserCVData extends Omit<UserData, 'uuid'>, Contacts {
  visibleCourses: VisibleCourses;
}

export interface UserDataToSubmit extends Omit<UserData, 'startFrom'> {
  startFrom: Dayjs;
  visibleCourses: VisibleCourses;
}

export interface AllDataToSubmit extends UserDataToSubmit, Contacts {}

export interface EditCVData extends AllUserCVData {
  expires: number | null;
  courses: CourseDataShortened[];
  visibleCourses: VisibleCourses;
}

export interface GetFullCVData extends AllUserCVData {
  expires: number | null;
  courses: CVStudentStats[];
  feedback: CVFeedback[];
}

export interface CVStudentStats {
  locationName: string;
  courseFullName: string;
  certificateId: string | null;
  isCourseCompleted: boolean;
  totalScore: number;
  rank: number | null;
  mentor: {
    githubId: string;
    name: string;
  };
}

export interface JobSeekerStudentStats extends CVStudentStats {
  courseName: string;
}

export interface JobSeekerData {
  name: string | null;
  desiredPosition: string | null;
  githubId: string;
  englishLevel: ResumeDtoEnglishLevelEnum | null;
  fullTime: boolean;
  locations: string | null;
  startFrom: string | null;
  courses: JobSeekerStudentStats[];
  feedback: JobSeekerFeedback[];
  expires: number;
  isHidden: boolean;
}

export interface CVFeedback {
  comment: string;
  feedbackDate: string;
}

export interface JobSeekerFeedback {
  badgeId: string;
}
