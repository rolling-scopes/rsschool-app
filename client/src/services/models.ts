import { Session } from 'components/withSession';
import {
  StudentBasic as CommonStudentBasic,
  DiscordServer,
  UserGroup,
  InterviewQuestion,
  InterviewQuestionCategory,
} from 'common/models';
import type { ProfileCourseDto } from 'api';

export interface Course extends ProfileCourseDto {}
export interface UpdateCourse extends ProfileCourseDto {
  disciplineId: number | null;
}

export interface StudentBasic extends CommonStudentBasic {}
export type { DiscordServer };
export type { UserGroup };
export type { InterviewQuestion };
export type { InterviewQuestionCategory };

export interface Mentor {
  lastName: string;
  firstName: string;
  githubId: string;

  studentId: number;
  userId: number;
  courseId: number;
}

export interface PageWithModalState<T> {
  data: T[];
  modalData: Partial<T> | null;
  modalAction: 'update' | 'create';
}

export interface CoursePageProps {
  session: Session;
  course: Course;
  params: Record<string, string>;
}

export type CourseOnlyPageProps = {
  course: Course;
  params?: Record<string, string>;
};
