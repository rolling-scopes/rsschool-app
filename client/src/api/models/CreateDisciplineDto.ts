/* istanbul ignore file */
/* tslint:disable */
/* eslint-disable */

export type CreateDisciplineDto = {
    name: string;
}
