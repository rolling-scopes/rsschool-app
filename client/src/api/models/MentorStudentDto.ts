/* istanbul ignore file */
/* tslint:disable */
/* eslint-disable */

export type MentorStudentDto = {
    name: string;
    id: number;
    cityName: string;
    countryName: string;
    githubId: string;
    totalScore: number;
    rank: number;
    hasFeedback: boolean;
}
