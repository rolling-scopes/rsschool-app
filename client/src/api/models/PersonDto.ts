/* istanbul ignore file */
/* tslint:disable */
/* eslint-disable */

export type PersonDto = {
    name: string;
    id: number;
}
