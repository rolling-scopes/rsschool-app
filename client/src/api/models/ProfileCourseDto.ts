/* istanbul ignore file */
/* tslint:disable */
/* eslint-disable */

export type ProfileCourseDto = {
    id: number;
    name: string;
    alias: string;
}
