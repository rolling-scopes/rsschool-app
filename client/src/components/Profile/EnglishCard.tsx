import * as React from 'react';
import isEqual from 'lodash/isEqual';
import { Typography, Select } from 'antd';
import CommonCard from './CommonCard';
import { GeneralInfo, ConfigurableProfilePermissions } from 'common/models/profile';
import { EnglishLevel } from 'common/models/index';
import { ChangedPermissionsSettings } from 'pages/profile';
import { CheckboxChangeEvent } from 'antd/lib/checkbox';

const { Text } = Typography;
const { Option } = Select;

import { TagOutlined } from '@ant-design/icons';

const ENGLISH_LEVELS: EnglishLevel[] = ['a0', 'a1', 'a1+', 'a2', 'a2+', 'b1', 'b1+', 'b2', 'b2+', 'c1', 'c1+', 'c2'];

type Props = {
  data: GeneralInfo;
  isEditingModeEnabled: boolean;
  permissionsSettings?: ConfigurableProfilePermissions;
  onPermissionsSettingsChange: (event: CheckboxChangeEvent, settings: ChangedPermissionsSettings) => void;
  onProfileSettingsChange: (event: any, path: string) => void;
};

class EnglishCard extends React.Component<Props> {
  private filterPermissions = ({ isEnglishVisible }: Partial<ConfigurableProfilePermissions>) => ({
    isEnglishVisible,
  });

  shouldComponentUpdate = (nextProps: Props) =>
    !isEqual(nextProps.data.englishLevel, this.props.data.englishLevel) ||
    !isEqual(nextProps.permissionsSettings?.isEnglishVisible, this.props.permissionsSettings?.isEnglishVisible) ||
    !isEqual(nextProps.isEditingModeEnabled, this.props.isEditingModeEnabled);

  render() {
    const { isEditingModeEnabled, permissionsSettings, onPermissionsSettingsChange, onProfileSettingsChange } =
      this.props;
    const { englishLevel } = this.props.data;
    return (
      <CommonCard
        title="Estimated English level"
        icon={<TagOutlined />}
        content={
          englishLevel ? (
            <Text style={{ textTransform: 'capitalize', fontSize: '48px' }}>{englishLevel}</Text>
          ) : undefined
        }
        noDataDescrption="English level isn't choosen"
        permissionsSettings={permissionsSettings ? this.filterPermissions(permissionsSettings) : undefined}
        isEditingModeEnabled={isEditingModeEnabled}
        onPermissionsSettingsChange={onPermissionsSettingsChange}
        profileSettingsContent={
          <div>
            <p style={{ fontSize: 18, marginBottom: 5 }}>
              <Text strong>English level:</Text>
            </p>
            <div style={{ marginBottom: 20 }}>
              <Select
                style={{ width: '100%' }}
                defaultValue={englishLevel || 'a0'}
                onChange={(event: any) => onProfileSettingsChange(event, 'generalInfo.englishLevel')}
              >
                {ENGLISH_LEVELS.map(level => (
                  <Option key={`settings-english-level-${level}`} value={level}>
                    {level.toUpperCase()}
                  </Option>
                ))}
              </Select>
            </div>
          </div>
        }
      />
    );
  }
}

export default EnglishCard;
