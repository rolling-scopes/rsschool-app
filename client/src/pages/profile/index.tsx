import * as React from 'react';
import get from 'lodash/get';
import set from 'lodash/set';
import isEqual from 'lodash/isEqual';
import cloneDeep from 'lodash/cloneDeep';
import mapValues from 'lodash/mapValues';
import clone from 'lodash/clone';
import pullAt from 'lodash/pullAt';
import { Result, Spin, message } from 'antd';
import Masonry from 'react-masonry-css';
import { Header } from 'components/Header';
import { NextRouter, withRouter } from 'next/router';
import { LoadingScreen } from 'components/LoadingScreen';
import withSession, { Session } from 'components/withSession';
import { UserService } from 'services/user';
import { CoursesService } from 'services/courses';
import {
  ProfileInfo,
  StudentStats,
  ConfigurableProfilePermissions,
  Contacts,
  GeneralInfo,
  Consent,
  Discord,
} from 'common/models/profile';

import MainCard from 'components/Profile/MainCard';
import AboutCard from 'components/Profile/AboutCard';
import DiscordCard from 'components/Profile/DiscordCard';
// import EnglishCard from 'components/Profile/EnglishCard';
import EducationCard from 'components/Profile/EducationCard';
import ContactsCard from 'components/Profile/ContactsCard';
import PublicFeedbackCard from 'components/Profile/PublicFeedbackCard';
import StudentStatsCard from 'components/Profile/StudentStatsCard';
import MentorStatsCard from 'components/Profile/MentorStatsCard';
import CoreJsIviewsCard from 'components/Profile/CoreJsIviewsCard';
import { CoreJsInterviewsData } from 'components/Profile/CoreJsIviewsCard';
import PreScreeningIviewCard from 'components/Profile/PreScreeningIviewCard';
import { withGoogleMaps } from 'components/withGoogleMaps';

import { CheckboxChangeEvent } from 'antd/lib/checkbox';
import { checkIsProfileOwner } from 'utils/profile-check';

type Props = {
  router: NextRouter;
  session: Session;
};

type State = {
  profile: ProfileInfo | null;
  initialPermissionsSettings: ConfigurableProfilePermissions | null;
  initialProfileSettings: ProfileInfo | null;
  isProfileOwner: boolean;
  isLoading: boolean;
  isSaving: boolean;
  isEditingModeEnabled: boolean;
  isInitialPermissionsSettingsChanged: boolean;
  isInitialProfileSettingsChanged: boolean;
};

export type ChangedPermissionsSettings = {
  permissionName: string;
  role: string;
};
export class ProfilePage extends React.Component<Props, State> {
  state: State = {
    profile: null,
    initialPermissionsSettings: null,
    initialProfileSettings: null,
    isProfileOwner: false,
    isLoading: true,
    isSaving: false,
    isEditingModeEnabled: false,
    isInitialPermissionsSettingsChanged: false,
    isInitialProfileSettingsChanged: false,
  };

  private onPermissionsSettingsChange = async (
    event: CheckboxChangeEvent,
    { permissionName, role }: ChangedPermissionsSettings,
  ) => {
    const { profile, initialPermissionsSettings } = this.state;
    const { checked } = event.target;

    if (profile?.permissionsSettings) {
      let changed = clone(get(profile.permissionsSettings, permissionName));

      if (role === 'all') {
        changed = mapValues(changed, () => checked);
      } else {
        changed[role] = checked;
      }

      const newPermissionsSettings = {
        ...profile.permissionsSettings,
        [permissionName]: changed,
      };
      const isInitialPermissionsSettingsChanged = !isEqual(newPermissionsSettings, initialPermissionsSettings);

      await this.setState({
        profile: {
          ...profile,
          permissionsSettings: newPermissionsSettings,
        },
        isInitialPermissionsSettingsChanged,
      });
    }
  };

  private onProfileSettingsChange = async (event: any = {}, path: string) => {
    const { profile, initialProfileSettings } = this.state;

    if (profile) {
      const newProfile = cloneDeep(profile);

      let isInitialProfileSettingsChanged;
      switch (path) {
        case 'generalInfo.location': {
          const countryName = (event && event.countryName) || profile?.generalInfo?.location.countryName;
          const cityName = (event && event.cityName) || profile?.generalInfo?.location.cityName;
          set(newProfile, `${path}.countryName`, countryName);
          set(newProfile, `${path}.cityName`, cityName);
          isInitialProfileSettingsChanged =
            initialProfileSettings?.generalInfo?.location?.cityName !== cityName ||
            initialProfileSettings?.generalInfo?.location?.countryName !== countryName;
          break;
        }
        case 'generalInfo.englishLevel': {
          set(newProfile, path, event);
          isInitialProfileSettingsChanged = initialProfileSettings?.generalInfo?.englishLevel !== event;
          break;
        }
        case 'generalInfo.educationHistory': {
          if (event.type === 'add') {
            newProfile.generalInfo?.educationHistory.push({
              graduationYear: null,
              faculty: null,
              university: null,
            });
          } else if (event.type === 'delete') {
            pullAt(newProfile.generalInfo?.educationHistory, [event.index]);
          }
          isInitialProfileSettingsChanged = !isEqual(
            initialProfileSettings?.generalInfo?.educationHistory,
            newProfile.generalInfo?.educationHistory,
          );
          break;
        }
        default: {
          set(newProfile, path, event.target.value);
          isInitialProfileSettingsChanged = get(newProfile, path) !== get(initialProfileSettings, path);
        }
      }

      await this.setState({ profile: newProfile, isInitialProfileSettingsChanged });
    }
  };

  private userService = new UserService();
  private coursesService = new CoursesService();

  private getCoursesInfo = async (profile: ProfileInfo) =>
    profile?.studentStats
      ? await Promise.all(profile?.studentStats?.map(({ courseId }) => this.coursesService.getCourse(courseId)))
      : [];

  private hadStudentCoreJSInterview = (stats: StudentStats[]) =>
    stats.some((student: StudentStats) => student.tasks.some(({ interviewFormAnswers }) => interviewFormAnswers));

  private getStudentCoreJSInterviews = (stats: StudentStats[]) =>
    stats
      .filter((student: StudentStats) => student.tasks.some(({ interviewFormAnswers }) => interviewFormAnswers))
      .map(({ tasks, courseFullName, courseName, locationName }) => ({
        courseFullName,
        courseName,
        locationName,
        interviews: tasks
          .filter(({ interviewFormAnswers }) => interviewFormAnswers)
          .map(({ interviewFormAnswers, score, comment, interviewer, name }) => ({
            score,
            comment,
            interviewer,
            answers: interviewFormAnswers,
            name,
          })),
      })) as CoreJsInterviewsData[];

  private fetchData = async () => {
    this.setState({ isLoading: true });

    const { router } = this.props;

    try {
      const githubId = router.query ? (router.query.githubId as string) : undefined;
      const profile = await this.userService.getProfileInfo(githubId);

      const coursesInfo = await this.getCoursesInfo(profile);

      const updateProfile = {
        ...profile,
        studentStats: profile.studentStats?.map(stats => ({
          ...stats,
          isCourseCompleted: coursesInfo.find(course => course.id === stats.courseId)?.completed ?? false,
        })),
      };

      let isProfileOwner = false;
      if (profile) {
        const userId = this.props.session.githubId;
        const profileId = profile.generalInfo!.githubId;
        isProfileOwner = checkIsProfileOwner(userId, profileId);
      }
      const initialPermissionsSettings = profile.permissionsSettings ? cloneDeep(profile.permissionsSettings) : null;
      const initialProfileSettings = profile ? cloneDeep(profile) : null;
      const isEditingModeEnabled = Boolean(router.asPath.match(/#edit/));

      this.setState({
        isLoading: false,
        profile: updateProfile,
        isProfileOwner,
        initialPermissionsSettings,
        isEditingModeEnabled,
        initialProfileSettings,
      });
    } catch (e) {
      this.setState({
        isLoading: false,
        profile: null,
        initialPermissionsSettings: null,
        initialProfileSettings: null,
      });
    }
  };

  private changeProfilePageMode = (mode: 'edit' | 'view') => {
    this.setState({ isEditingModeEnabled: mode === 'edit' ? true : false });
  };

  private onSaveSuccess() {
    message.success('Profile was successesfully saved');
  }

  private onSaveError() {
    message.error('Error has occured. Please check your connection and try again');
  }

  private saveProfile = async () => {
    const { profile, isInitialPermissionsSettingsChanged, isInitialProfileSettingsChanged } = this.state;

    this.setState({ isSaving: true });

    if (profile) {
      try {
        const { permissionsSettings, generalInfo, contacts, consents, discord } = profile;
        await this.userService.saveProfileInfo({
          permissionsSettings: permissionsSettings as ConfigurableProfilePermissions,
          generalInfo: generalInfo as GeneralInfo,
          contacts: contacts as Contacts,
          consents: consents as Consent[],
          discord: discord as Discord,
          isPermissionsSettingsChanged: isInitialPermissionsSettingsChanged,
          isProfileSettingsChanged: isInitialProfileSettingsChanged,
        });

        const initialPermissionsSettings = permissionsSettings ? cloneDeep(permissionsSettings) : null;
        const initialProfileSettings = profile ? cloneDeep(profile) : null;
        this.setState({
          isSaving: false,
          initialPermissionsSettings,
          initialProfileSettings,
          isInitialPermissionsSettingsChanged: false,
          isInitialProfileSettingsChanged: false,
        });
        this.onSaveSuccess();
      } catch (e) {
        this.setState({ isSaving: false });
        this.onSaveError();
      }
    }
  };

  authorizeDiscord = async () => {
    await this.setState({ isLoading: true });

    const discord = await this.userService.getDiscordIds();

    if (discord) {
      this.setState(({ profile, ...state }) => ({
        ...state,
        profile: {
          ...profile,
          discord,
        },
        isInitialProfileSettingsChanged: true,
      }));

      await this.saveProfile();
      this.props.router.replace('/profile');
    }

    await this.setState({ isLoading: false });
  };

  componentDidMount() {
    // it's a dirty hack to fix an issue with empty query params
    // see: https://nextjs.org/docs/routing/dynamic-routes#caveats
    //
    // >> After hydration, Next.js will trigger an update to your application
    // >> to provide the route parameters in the query object.
    setTimeout(async () => {
      await this.fetchData();
      await this.authorizeDiscord();
    }, 100);
  }

  render() {
    const {
      profile,
      isEditingModeEnabled,
      initialPermissionsSettings,
      isInitialPermissionsSettingsChanged,
      isInitialProfileSettingsChanged,
      isProfileOwner,
    } = this.state;
    const isEditingModeVisible = initialPermissionsSettings && isEditingModeEnabled ? isEditingModeEnabled : false;
    const isSaveButtonVisible = isInitialPermissionsSettingsChanged || isInitialProfileSettingsChanged;

    const cards = [
      profile?.generalInfo && (
        <MainCard
          data={profile.generalInfo}
          isEditingModeEnabled={isEditingModeVisible}
          permissionsSettings={profile.permissionsSettings}
          onPermissionsSettingsChange={this.onPermissionsSettingsChange}
          onProfileSettingsChange={this.onProfileSettingsChange}
        />
      ),
      profile?.generalInfo?.aboutMyself !== undefined && (
        <AboutCard
          data={profile.generalInfo}
          isEditingModeEnabled={isEditingModeVisible}
          permissionsSettings={profile.permissionsSettings}
          onPermissionsSettingsChange={this.onPermissionsSettingsChange}
          onProfileSettingsChange={this.onProfileSettingsChange}
        />
      ),
      // TODO: Temporary disabled
      // profile?.generalInfo?.englishLevel !== undefined && (
      //   <EnglishCard
      //     data={profile.generalInfo}
      //     isEditingModeEnabled={isEditingModeVisible}
      //     permissionsSettings={profile.permissionsSettings}
      //     onPermissionsSettingsChange={this.onPermissionsSettingsChange}
      //     onProfileSettingsChange={this.onProfileSettingsChange}
      //   />
      // ),
      profile?.generalInfo?.educationHistory !== undefined && (
        <EducationCard
          data={profile.generalInfo}
          isEditingModeEnabled={isEditingModeVisible}
          permissionsSettings={profile.permissionsSettings}
          onPermissionsSettingsChange={this.onPermissionsSettingsChange}
          onProfileSettingsChange={this.onProfileSettingsChange}
        />
      ),
      profile?.contacts !== undefined && (
        <ContactsCard
          data={profile.contacts}
          isEditingModeEnabled={isEditingModeVisible}
          permissionsSettings={profile.permissionsSettings}
          onPermissionsSettingsChange={this.onPermissionsSettingsChange}
          onProfileSettingsChange={this.onProfileSettingsChange}
        />
      ),
      profile?.discord !== undefined && <DiscordCard data={profile.discord} isProfileOwner={isProfileOwner} />,
      profile?.publicFeedback?.length && (
        <PublicFeedbackCard
          data={profile.publicFeedback}
          isEditingModeEnabled={isEditingModeVisible}
          permissionsSettings={profile.permissionsSettings}
          onPermissionsSettingsChange={this.onPermissionsSettingsChange}
        />
      ),
      profile?.studentStats?.length && (
        <StudentStatsCard
          username={this.props.session.githubId}
          data={profile.studentStats}
          isProfileOwner={isProfileOwner}
          isEditingModeEnabled={isEditingModeVisible}
          permissionsSettings={profile.permissionsSettings}
          onPermissionsSettingsChange={this.onPermissionsSettingsChange}
        />
      ),
      profile?.mentorStats?.length && (
        <MentorStatsCard
          data={profile.mentorStats}
          isEditingModeEnabled={isEditingModeVisible}
          permissionsSettings={profile.permissionsSettings}
          onPermissionsSettingsChange={this.onPermissionsSettingsChange}
        />
      ),
      profile?.studentStats?.length && this.hadStudentCoreJSInterview(profile.studentStats) && (
        <CoreJsIviewsCard data={this.getStudentCoreJSInterviews(profile.studentStats)} />
      ),
      profile?.stageInterviewFeedback?.length && <PreScreeningIviewCard data={profile.stageInterviewFeedback} />,
    ].filter(Boolean) as JSX.Element[];

    return (
      <>
        <LoadingScreen show={this.state.isLoading}>
          <Header
            username={this.props.session.githubId}
            onChangeProfilePageMode={this.changeProfilePageMode}
            isProfilePage={initialPermissionsSettings ? true : false}
            isProfileEditingModeEnabled={isEditingModeVisible}
            isSaveButtonVisible={isSaveButtonVisible}
            onSaveClick={this.saveProfile}
          />
          <Spin spinning={this.state.isSaving} delay={200}>
            {this.state.profile ? (
              <div style={{ padding: 10 }}>
                <Masonry
                  breakpointCols={{
                    default: 4,
                    1100: 3,
                    700: 2,
                    500: 1,
                  }}
                  className="masonry"
                  columnClassName="masonry-column"
                >
                  {cards.map((card, idx) => (
                    <div style={{ marginBottom: 16 }} key={`card-${idx}`}>
                      {card}
                    </div>
                  ))}
                </Masonry>
                <style jsx global>{`
                  .masonry {
                    display: flex;
                    margin-left: -16px;
                    width: auto;
                  }
                `}</style>
                <style jsx global>{`
                  .masonry-column {
                    padding-left: 16px;
                    background-clip: padding-box;
                  }
                `}</style>
              </div>
            ) : (
              <>
                <Result status={'403' as any} title="No access or user does not exist" />
              </>
            )}
          </Spin>
        </LoadingScreen>
      </>
    );
  }
}

export default withGoogleMaps(withRouter(withSession(ProfilePage)));
