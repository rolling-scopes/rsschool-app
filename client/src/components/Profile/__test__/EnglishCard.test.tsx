import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import EnglishCard from '../EnglishCard';
import { GeneralInfo } from 'common/models/profile';

describe('EnglishCard', () => {
  describe('Should render correctly', () => {
    it('if "englishLevel" is present', () => {
      const output = shallow(
        <EnglishCard
          data={
            {
              englishLevel: 'a2',
            } as GeneralInfo
          }
          isEditingModeEnabled={false}
          onPermissionsSettingsChange={jest.fn()}
          onProfileSettingsChange={jest.fn()}
        />,
      );
      expect(shallowToJson(output)).toMatchSnapshot();
    });
    it('if "aboutMyself" is not present', () => {
      const output = shallow(
        <EnglishCard
          data={{} as GeneralInfo}
          isEditingModeEnabled={false}
          onPermissionsSettingsChange={jest.fn()}
          onProfileSettingsChange={jest.fn()}
        />,
      );
      expect(shallowToJson(output)).toMatchSnapshot();
    });
  });
  describe('filterPermissions', () => {
    it('should left only "isAboutVisible" in "permissionsSettings" object', () => {
      const permissionsSettings = {
        isProfileVisible: { all: true },
        isAboutVisible: { all: true, mentor: true, student: true },
        isEducationVisible: { all: true, mentor: true, student: true },
        isEnglishVisible: { all: false, student: false },
        isEmailVisible: { all: true, student: true },
        isTelegramVisible: { all: false, student: false },
        isSkypeVisible: { all: true, student: true },
        isPhoneVisible: { all: false, student: false },
        isContactsNotesVisible: { all: true, student: true },
        isLinkedInVisible: { all: false, mentor: false, student: false },
        isPublicFeedbackVisible: { all: true, mentor: true, student: true },
        isMentorStatsVisible: { all: true, mentor: true, student: true },
        isStudentStatsVisible: { all: true, student: true },
      };
      const wrapper = shallow(
        <EnglishCard
          data={{} as GeneralInfo}
          isEditingModeEnabled={false}
          permissionsSettings={permissionsSettings}
          onPermissionsSettingsChange={jest.fn()}
          onProfileSettingsChange={jest.fn()}
        />,
      );
      const instance: any = wrapper.instance();
      const result = instance.filterPermissions(permissionsSettings);
      expect(result).toEqual({
        isEnglishVisible: { all: false, student: false },
      });
    });
  });
});
