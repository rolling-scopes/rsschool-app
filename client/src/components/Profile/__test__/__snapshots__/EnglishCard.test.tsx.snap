// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EnglishCard Should render correctly if "aboutMyself" is not present 1`] = `
<div>
  <div
    class="ant-card ant-card-bordered"
  >
    <div
      class="ant-card-head"
    >
      <div
        class="ant-card-head-wrapper"
      >
        <div
          class="ant-card-head-title"
        >
          <h2
            class="ant-typography ant-typography-ellipsis ant-typography-single-line ant-typography-ellipsis-single-line"
            style="font-size: 16px; margin-bottom: 0px;"
          >
            <span
              aria-label="tag"
              class="anticon anticon-tag"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="tag"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M938 458.8l-29.6-312.6c-1.5-16.2-14.4-29-30.6-30.6L565.2 86h-.4c-3.2 0-5.7 1-7.6 2.9L88.9 557.2a9.96 9.96 0 000 14.1l363.8 363.8c1.9 1.9 4.4 2.9 7.1 2.9s5.2-1 7.1-2.9l468.3-468.3c2-2.1 3-5 2.8-8zM459.7 834.7L189.3 564.3 589 164.6 836 188l23.4 247-399.7 399.7zM680 256c-48.5 0-88 39.5-88 88s39.5 88 88 88 88-39.5 88-88-39.5-88-88-88zm0 120c-17.7 0-32-14.3-32-32s14.3-32 32-32 32 14.3 32 32-14.3 32-32 32z"
                />
              </svg>
            </span>
             
            Estimated English level
          </h2>
        </div>
      </div>
    </div>
    <div
      class="ant-card-body"
    >
      <div
        class="ant-empty ant-empty-normal"
      >
        <div
          class="ant-empty-image"
        >
          <svg
            class="ant-empty-img-simple"
            height="41"
            viewBox="0 0 64 41"
            width="64"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              fill="none"
              fill-rule="evenodd"
              transform="translate(0 1)"
            >
              <ellipse
                class="ant-empty-img-simple-ellipse"
                cx="32"
                cy="33"
                rx="32"
                ry="7"
              />
              <g
                class="ant-empty-img-simple-g"
                fill-rule="nonzero"
              >
                <path
                  d="M55 12.76L44.854 1.258C44.367.474 43.656 0 42.907 0H21.093c-.749 0-1.46.474-1.947 1.257L9 12.761V22h46v-9.24z"
                />
                <path
                  class="ant-empty-img-simple-path"
                  d="M41.613 15.931c0-1.605.994-2.93 2.227-2.931H55v18.137C55 33.26 53.68 35 52.05 35h-40.1C10.32 35 9 33.259 9 31.137V13h11.16c1.233 0 2.227 1.323 2.227 2.928v.022c0 1.605 1.005 2.901 2.237 2.901h14.752c1.232 0 2.237-1.308 2.237-2.913v-.007z"
                />
              </g>
            </g>
          </svg>
        </div>
        <div
          class="ant-empty-description"
        >
          English level isn't choosen
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`EnglishCard Should render correctly if "englishLevel" is present 1`] = `
<div>
  <div
    class="ant-card ant-card-bordered"
  >
    <div
      class="ant-card-head"
    >
      <div
        class="ant-card-head-wrapper"
      >
        <div
          class="ant-card-head-title"
        >
          <h2
            class="ant-typography ant-typography-ellipsis ant-typography-single-line ant-typography-ellipsis-single-line"
            style="font-size: 16px; margin-bottom: 0px;"
          >
            <span
              aria-label="tag"
              class="anticon anticon-tag"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="tag"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M938 458.8l-29.6-312.6c-1.5-16.2-14.4-29-30.6-30.6L565.2 86h-.4c-3.2 0-5.7 1-7.6 2.9L88.9 557.2a9.96 9.96 0 000 14.1l363.8 363.8c1.9 1.9 4.4 2.9 7.1 2.9s5.2-1 7.1-2.9l468.3-468.3c2-2.1 3-5 2.8-8zM459.7 834.7L189.3 564.3 589 164.6 836 188l23.4 247-399.7 399.7zM680 256c-48.5 0-88 39.5-88 88s39.5 88 88 88 88-39.5 88-88-39.5-88-88-88zm0 120c-17.7 0-32-14.3-32-32s14.3-32 32-32 32 14.3 32 32-14.3 32-32 32z"
                />
              </svg>
            </span>
             
            Estimated English level
          </h2>
        </div>
      </div>
    </div>
    <div
      class="ant-card-body"
    >
      <span
        class="ant-typography"
        style="text-transform: capitalize; font-size: 48px;"
      >
        a2
      </span>
    </div>
  </div>
</div>
`;
